src/internal/storage/btree/Btree.h

```diff
@@ -0,0 +1,776 @@
+		auto &as_branch = branch();
+		as_branch.refs_[it.index()] = ref;
+		++m_numfilled;
+
+		if (is_balanced())
+			return it;
+
+		return rebalance(SmallestRefAction::MOVE);
+	}
+
+	std::pair<Iterator, bool> find(const Key &target) noexcept {
+		auto [lo, _] = range();
+		auto hi = lo + m_numfilled;
+		const auto beginning = lo;
+
+		while (lo < hi) {
+			auto mid = lo;
+			std::advance(mid, std::distance(lo, hi) / 2);
+			if (target > *mid)
+				lo = mid + 1;
+			else
+				hi = mid;
+		}
+
+		long idx = std::distance(beginning, hi);
+		auto it = Iterator{*this, idx, *m_tree};
+		return std::make_pair(it, *hi == target && m_numfilled > 0);
+	}
+
+private:
+	Header m_header;
+	Bt *m_tree;
+
+	std::variant<LeafMeta, BranchMeta> m_meta;
+	long m_numfilled{0};
+};
+
+template<BtreeConfig Config>
+class Btree<Config>::Iterator final {
+public:// Iterator traits
+	using difference_type = long;
+	using value_type = Node;
+	using pointer = Node *;
+	using reference = Node &;
+	using iterator_category = std::bidirectional_iterator_tag;
+
+private:
+	using Val = Config::Val;
+	using Key = Config::Key;
+
+public:// Constructor
+	Iterator() = default;
+
+	Iterator(const Iterator &) noexcept = default;
+
+	/// Tree::begin iterator
+	explicit Iterator(Btree &tree) : m_node{nullptr}, m_index{0}, m_tree{&tree} {
+		Node *cur = tree.root_ptr();
+		while (cur->is_branch()) {
+			const auto &branch_meta = cur->branch();
+
+			assert(cur->numfilled() > 0);
+			const Position last_link = branch_meta.links_[0];
+
+			cur = &Node::Cache::the().fetch(last_link);
+		}
+
+		m_node = cur;
+	}
+
+	/// Iterator to a specific location in the tree
+	Iterator(Node &node, long index, Btree &tree)
+	    : m_node{&node}, m_index{index}, m_tree{&tree} {}
+
+	static Iterator begin(Btree &tree) { return Iterator{tree}; }
+
+	static Iterator end(Btree &tree) {
+		Node *cur = tree.root_ptr();
+		while (cur->is_branch()) {
+			const auto &branch_meta = cur->branch();
+
+			assert(cur->numfilled() > 0);
+			const Position last_link = branch_meta.links_[cur->numfilled() - 1];
+
+			cur = &Node::Cache::the().fetch(last_link);
+		}
+
+		return Iterator{*cur, cur->numfilled(), tree};
+	}
+
+public:
+	~Iterator() noexcept = default;
+
+	Iterator &operator=(const Iterator &) = default;
+
+public:// Properties
+	Btree<Config> &tree() const noexcept { return *m_node->m_tree; }
+
+	optional_cref<Val> val() const noexcept {
+		assert(m_node->numfilled() > m_index);
+		assert(m_tree && m_node && m_index >= 0);
+		if (m_node->is_branch())
+			return std::nullopt;
+		const auto &leaf_meta = m_node->leaf();
+		return std::make_optional(std::cref(leaf_meta.vals_[m_index]));
+	}
+
+	optional_cref<Key> key() const noexcept {
+		assert(m_node->numfilled() > m_index);
+		assert(m_tree && m_node && m_index >= 0);
+
+		return std::make_optional(std::cref(m_node->at(m_index)));
+	}
+
+	const Node &node() const noexcept { return *m_node; }
+	Node &node_mut() noexcept { return *m_node; }
```

Comment:
This is also not used and I didn't understand the name node_mut and for what it will be usable. Probably remove?

Reply:
`_mut` stands for mutable - it returns a reference which enables the caller to modify the value of the node. Example:
```
it.node_mut().function_which_requires_this_to_be_mutable();
```

I am unsure whether to remove it. I agree that it is not used, **but** I think it should remain with the addition of `[maybe_unused]` flag since it is part of the public API and I expect it will be useful, since there is no other function which performs same or even similar actions.

What do you think?